Add tutorial for generating time series data using node and node-postgres

autophagy opened the pull request:
I tried to tack as close as possible to the Python tutorial, as it had a good structure and flow. 

@mechanomi I have a question for you, specifically the:

```
    > client.query('SELECT * FROM load ORDER BY timestamp DESC', (err, res) => {
    ...   if (err) {
    ...       console.log(err.stack)
    ...     } else {
    ...       console.log(res.rows)
    ...     }
    ... })
```

Example. Previously, we just used ``await`` for async ones in the console. It could be rewritten as:

```
let result = await client.query('SELECT * FROM load ORDER BY timestamp DESC')
console.log(result.rows)
```

This doesn't have any exception handling, so wouldn't be best practice, but since this is just an intro tutorial based on playing around in the CLI, perhaps that's not so important? What do you think?

autophagy committed:
Add tutorial for generating time series data using node and node-postgres
Co-authored-by: Naomi Slater <23469+mechanomi@users.noreply.github.com>

autophagy commented:
@mechanomi Also, bear in mind that `os.loadavg` returns only `0` for Windows users, since the notion of a load average is a UNIX thing. I think that's fine?

matthijskrul reviewed (commented):
Not Naomi but imho tutorials don't necessarily need exception handling, as they are intended to get a specific result with a specific method. Of course, you do need to be sure _yourself_ that the method works! 

Also, would it be something to rewrite "SELECT out of as a phrase" (112)? Seems a little too... shorthand.

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+
+CrateDB must be :ref:`installed and running <install-run>`.
+
+Make sure you're running up-to-date Node/npm.
```
can you separate out Node and npm?

e.g., on one line, say:

```
Make sure you're running up-to-date `Node`_  (we recommend X or higher)..
```

link the name "Node" to the Node homepage. and I'd include a recommended version

then for npm, how about including a line that shows you how to upgrade npm to the latest version using npm itself? you could then also add a small fix to the python tutorial that does the same with pip

(please also link the first instance of "npm" to the npm homepage)

autophagy replied:
I honestly have no idea what would be the recommended version here. I have the latest installed but I didn't try the tutorial out on all the other node versions.

nomicode replied:
perhaps just this then:

```
Make sure you're running up-to-date `Node`_
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+Make sure you're running up-to-date Node/npm.
+
+The `node-postgres`_ driver is our officially supported driver for Javascript.
+With this in mind, use `npm`_ to install the `pg`_ library:
```
I think we can probably drop the bit about officially supporting it. that can probably be mentioned on the list of supported clients page. in this context, it's implicit, I think

how about something like this:

```
Use npm to install the `node-postgres`_ library:

[...]
```

here, I've used `node-postgres` in full in the prose, because it's easier to read and understand, but then switched to `pg` in the command, which hopefully is enough of an explanation for the reader that these two things are synonyms

thoughts?

autophagy replied:
I think this makes sense, aye.

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+
+    $ npm install pg
+
+The rest of this tutorial is designed for Node.js's `interactive REPL mode`_ so that
```
I think you can say "Node's". I realize "Node.js" is the full name, but we only really need to use the full name in the first instance we use it  (which you already have done)

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+    [ 0.9794921875, 1.08740234375, 0.97314453125 ]
+
+
+The function returns a triple of `numbers`_ corresponding the average system load over the
```
missing a "to" before the "the average"

can you switch to using the phrasing used for the python tutorial, which includes spelling out the numbers under 10 (which is our house style), e.g.:

"... corresponding to the one minute average, five-minute average, and 15-minute average."

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+
+    > const { Client } = require('pg')
+
+Then, `connect`_ to CrateDB, using the port for the `Postgres Wire Protocol`_::
```
```
, using the port for the `Postgres Wire Protocol`_
```

is this bit necessary? it seems to me that it would be implied. is the `port` argument even necessary in the command below?

autophagy replied:
I suppose not - I just thought it's better to be explicit here than implicit, since the other tutorials use `4200`. I can remove it if you think its necessary?

I'd prefer to keep the port argument in the connection string though, again, for the sake of explicitness.

nomicode replied:
hmm. I see. why do the other tutorials use 4200 (besides the obvious answer that that's how I wrote them lol)? I went with the defaults in the docs for those clients. does that imply those clients not using the postgresql wire protocol?

autophagy replied:
Yep! The python client communicates over 4200, our HTTP port. Our own-written clients don't use the pg wire protocol.

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+    ... })
+    > await client.connect()
+
+The client can also be initialised with a single connection string, if prefered::
```
for these tutorials, I want to pick one single way to do things, to keep things simple. then for other options, I would like to link through to the docs so people can explore other ways to do it

what do you think?

using the connection string might be the better option here, as it's more concise, and matches the python tutorial

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+        avg_15m: 0.22802734 } ]
+
+Here we have recorded three sets of load averages with a corresponding
+timestamp. We can now close the connection::
```
make a new para:

```
When you're ready, you can close the connection::
```

this is in case the user wants to continue experimenting

however, I would also ask: is this necessary? would bad things happen if they don't close the connection? I realize it's probably good practice to do this in a real world app, but as this is a repl based tutorial, is it needed?

autophagy replied:
Yeah, i think its probably fine to drop it.

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+
+Now we have the basics figured out, let's automate the data collection.
+
+Our approach will change a little bit here. We were previously using a `Client`_ to
```
I suspect that "client" can be lowercase throughout this para

nomicode replied:
I'd also maybe experiment with starting a new para at "We were" and seeing how it affects the flow. short n sweet paras are good for keeping people's attention

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+they need to be recreated. However, creating a new Client requires a handshake with CrateDB, and
+this overhead cost can be prohibative if we are creating new clients rapidly.
+
+Instead, we can use a `Connection Pool`_ to manage our connections. Create a new file called
```
new para at "Create a new file"

nomicode replied:
would probably lowercase "connection pool", even though it's a class name, it also functions well as a generic noun phrase

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+Our approach will change a little bit here. We were previously using a `Client`_ to
+connect to and insert data into CrateDB. However, Clients are ephemeral, and once closed
+they need to be recreated. However, creating a new Client requires a handshake with CrateDB, and
+this overhead cost can be prohibative if we are creating new clients rapidly.
```
I'm curious

if you look at the python tutorial, you will notice that I create a new connection each time the main loop evals. I did this primarily to mimic the behavior of the shell script tutorial

but then it occurred to me that keeping open a long-running connection invites problems. if a network issue arises, that while loop is going to start producing errors

I also didn't want to complicate the script too much by introducing connection management

so I would ask, which of the following sounds better to you:

1. drop the connection pool stuff for node. and just try to recreate the connection each time, mimicking the existing two tutorials (and building in a little resiliency -- though not much, because we're still not doing exception handling)

2. same as (1) but add exception handling 

3. keeping the connection pool use here, and updating the python tutorial to do something similar

4. keeping the connection pool use here, and not updating the python tutorial

autophagy replied:
I would probably say 3. The crate-python driver already does pooling under the hood, see: https://github.com/crate/crate-python/blob/750fe0042c25b8c7a77ffb8b3afa096926f1fd8d/src/crate/client/http.py#L92 - when you do `connect` with crate-python, it gets a connection pool from urllib3 and manages the connection with that. With the postgres driver, we have to do that explicitly. In the python driver, a connection is actually a wrapper around a connection pool. It's probably not needed to create a new connection on every main loop of the python driver, actually.

nomicode replied:
cool okay. let's do that then

after we merge this pr, do you wanna submit a patch for the python tutorial as you have more familiarity with the client? (otherwise, happy to look into it myself!)

autophagy replied:
Will do!

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+        user: 'crate',
+        host: 'localhost',
+        database: 'doc',
+        port: 5432,
```
we should make sure this matches whichever connection method you choose as the one to recommend for this tutorial (assuming you agree with my comment above)

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+                } else {
+                    console.log("INSERT OK")
+                }
+            }
```
should we add a note about adding exception handling?

also: should we add this to the python example?

or, alternatively, should we strip this demo down so it's as basic as the python one with no exception handling at all. the downside of this is that the script is brittle for actual real-world use and we pass up the opportunity to introduce exception handling as a tool. the upside is: it keeps the tutorial simple and means that the end example isn't using too much extra code that hasn't been used in the repl sections

autophagy replied:
Hm, I'm torn. On the one hand, without error handling the above could be written as a much simpler:

```
await pool.query("INSERT INTO load (avg_1m, avg_5m, avg_15m) VALUES (?, ?, ?)", os.loadavg())
console.log("INSERT OK")
```

But, on the other hand, I would also be tetchy about a tutorial that says, explicitly, "INSERT OK" without actually ensuring that the insert was ok. I think adding a little bit of exception handling to the python section is probably the best thing to do here. What do you think?

nomicode replied:
okay yes, you've convinced me. :D let's keep it, and let's patch the python tutorial afterward. do you want to do that or should I?

autophagy replied:
I'll take care of it!

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+.. code-block:: console
+
+    $ node monitor-load.js
+    INSERT OK
```
I am missing the "CONNECT OK", which I added to python so that it mimics the crash tutorial. haha. but it's not necessary. and whether it should even be here or not is going to be answered by your answer to my above questions :)

autophagy replied:
I think letting the pool handle it is fine - if connection is wrong, the user will know. If the connection is fine, they will see some insert stuff. Not sure if its really necessary.

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+.. image:: ../_assets/img/generate-time-series/rows.png
+
+Lots of freshly generated time series data, ready for use.
+
```
please add an extra empty newline to separate off the list of links

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+.. _pg: https://www.npmjs.com/package/pg
+.. _SELECT: https://crate.io/docs/crate/reference/en/latest/general/dql/selects.html
+.. _system load: https://en.wikipedia.org/wiki/Load_(computing)
+.. _time series: https://en.wikipedia.org/wiki/Time_series
```
please sort this list of links alphabetically (case insensitive)

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,232 @@
+.. _time series: https://en.wikipedia.org/wiki/Time_series
+
+
+
```
please trim these trailing empty newlines down to one empty newline

nomicode commented:
> Also, would it be something to rewrite "SELECT out of as a phrase" (112)? Seems a little too... shorthand.

do you mean to suggest a rewrite to this sentence:

"When you’re done, you can SELECT that data back out of CrateDB, like so:"

if so, I think it's fine as is :) because it reads like a normal sentence, and capitalizing and linking the "SELECT" also clues the reader into its special meaning as an SQL verb

nomicode commented:
> Previously, we just used `await` for async ones in the console.

when you say "previously", what are you referring to?

what is the benefit of using async handling stuff in the repl? I'm a complete Node newb so this might be a very common way of interacting with things. but outside of the context of an actual application that is making use of async, it seems a little over-complex. thoughts?

I'd like to keep the examples/tutorial as simple as possible. but I also don't want to do that if it produces non-idiomatic code examples! and I don't have enough exp with Node to know what's what here

nomicode commented:
> @mechanomi Also, bear in mind that `os.loadavg` returns only `0` for Windows users, since the notion of a load average is a UNIX thing. I think that's fine?

argh. this is very frustrating haha. I originally wanted to write these tutorials to collect fan speeds, or CPU temp, or something like that, because it's much more like the sort of IoT smart sensor data that our customers might have in mind

but then I found out that macOS hides this information quite well. well enough that I'd have to write two separate collections of tutorials. one for Linux, with full access to a rich set of data, and a simplified one for macOS

I went with load averages bcus I naively thought they were a lowest common denominator

what do you think we should do here? is there a lowest common denominator that would work in place of load average? perhaps just something like memory usage, CPU utilization %

I'd rather have a slightly boring collection of tutorials that produce consistent outputs across all our supported operating systems than have to deal with the complexity of either (1) writing different collections for different operating systems, or (2) writing one collection, but having a bunch of "if you're on Linux, follow this section, otherwise, if you're on macOS, jump to section X. Windows users, go to section Y" or whatever

autophagy commented:
> > Previously, we just used `await` for async ones in the console.
> 
> when you say "previously", what are you referring to?
> 
> what is the benefit of using async handling stuff in the repl? I'm a complete Node newb so this might be a very common way of interacting with things. but outside of the context of an actual application that is making use of async, it seems a little over-complex. thoughts?
> 
> I'd like to keep the examples/tutorial as simple as possible. but I also don't want to do that if it produces non-idiomatic code examples! and I don't have enough exp with Node to know what's what here

Theres not so much benefit to using async handling in the repl - but the node-postgres driver works in a completely async way - each one of the calls like `connect`, `query` etc are asynchronous.

autophagy commented:
> I'd rather have a slightly boring collection of tutorials that produce consistent outputs across all our supported operating systems than have to deal with the complexity of either (1) writing different collections for different operating systems, or (2) writing one collection, but having a bunch of "if you're on Linux, follow this section, otherwise, if you're on macOS, jump to section X. Windows users, go to section Y" or whatever

Yeah, this is frustrating, and I'm not so sure how to address this. If we're using system metrics to generate sample data, its almost inevitable that the methods will differ between different OSes. Perhaps free memory could be an example used instead? Assuming that it works across all 3.

nomicode commented:
> Theres not so much benefit to using async handling in the repl - but the node-postgres driver works in a completely async way - each one of the calls like `connect`, `query` etc are asynchronous.

@autophagy in that case, I'm inclined to agree with how you've approached it. thank you for clarifying! perhaps you could add a note block explaining this in the tutorial?

nomicode commented:
> Yeah, this is frustrating, and I'm not so sure how to address this. If we're using system metrics to generate sample data, its almost inevitable that the methods will differ between different OSes. Perhaps free memory could be an example used instead? Assuming that it works across all 3.

@autophagy I've been thinking about this all day. I think free memory is a good choice. it's easy to understand and I expect it is accessible on all systems. and it doesn't come with the complications that CPU utilization comes with (multiple CPUs, etc)

can you rewrite this tutorial to use free memory? afterward, I will patch the existing tutorials to use free memory (unless you want to, ofc)

the only thing we need to do is standardize on a unit of measurement. I guess bytes is the most logical smallest possible unit?

autophagy commented:
@mechanomi Thank you for the feedback! 

1. Good idea, i'll add a note block for the async stuff.

2. Yeah, I think memory with bytes makes sense - but im not sure how to do this on mac. I dont believe it has the `free` command that I use with linux. We could also use `vmstat` which I think mac has an equivalent of called `vm_stat`. It's output on linux looks like:

```
❯ vmstat
procs -----------memory---------- ---swap-- -----io---- -system-- ------cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
 0  0      0 2036944 336664 7050480    0    0    18    55  137  283  7  3 89  1  0
```

However the output on mac looks like:

```
➜ vm_stat
Mach Virtual Memory Statistics: (page size of 4096 bytes)
Pages free:                              139824.
Pages active:                           1673261.
Pages inactive:                         1073096.
Pages speculative:                         2909.
Pages throttled:                              0.
Pages wired down:                        788478.
Pages purgeable:                         226206.
[...]
```

So it's not exactly very similar across platforms. The only other thing I can think of is scraping some kind of data from the web using `curl` :sweat_smile:

nomicode commented:
ooh ooh, let's do this:

http://open-notify.org/Open-Notify-API/ISS-Location-Now/

nomicode commented:
or this maybe: https://wheretheiss.at/

https://api.wheretheiss.at/v1/satellites/25544

nomicode commented:
@autophagy I think we should go with the open notify one. even though it provides less data, it looks to be more recently maintained. also, they published the source code, so we can fork a copy, and work on hosting our own

https://github.com/open-notify

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+.. _gen-ts-javascript:
+
+==========================================
+Generate time series data using Javascript
```
```suggestion
Generate time series data using JavaScript
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+==========================================
+
+This tutorial will show you how to generate some :ref:`experimental time series
+data <gen-ts>` by consuming telemetry data from the `International Space Station`_
```
maybe "[...] how to generate some experimental time series data from information about the International Space Station (ISS) [...]"

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+
+This tutorial will show you how to generate some :ref:`experimental time series
+data <gen-ts>` by consuming telemetry data from the `International Space Station`_
+using Javascript, specifically `Node.js`_.
```
```suggestion
using Javascript (specifically, `Node.js`_).
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+
+Make sure you're running up-to-date `Node`_.
+
+Also make sure you're running up-to-date `npm`_. To upgrade npm:
```
how about:

```
Make sure you're running an up-to-date `Node`_ installation.

Then, upgrade to the latest `npm`_ version:
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+
+    $ npm install -g npm@latest
+
+Use npm to install the `node-postgres`_ library:
```
```suggestion
Install the `node-postgres`_ and `Axios`_ libraries:
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+
+.. code-block:: console
+
+    $ npm install pg
```
```suggestion
    $ npm install pg axios
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+.. code-block:: console
+
+    $ npm install axios
+
```
delete this? I've integrated it into the command above for brevity

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+    $ npm install axios
+
+
+Both the ``axios`` and ``node-postgres`` libraries operate using `Promises`_. Promises are
```
```suggestion
The ``node-postgres`` and ``axios` libraries both use `Promises`_. Promises are
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+Both the ``axios`` and ``node-postgres`` libraries operate using `Promises`_. Promises are
+a way of encapsulating the eventual result of an asynchronous operation. Both ``axios``, when
+retrieving content from a URL, and ``node-postgres``, when connecting and querying a database,
+use Promises as a way of asynchronously doing those operations.
```
would delete the bit starting "Both"

if you want to include info about what they use promises for you could reword the first sentence more like this:

"The ``node-postgres`` and ``axios` libraries both use `Promises`_ when performing network operations."

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+retrieving content from a URL, and ``node-postgres``, when connecting and querying a database,
+use Promises as a way of asynchronously doing those operations.
+
+If you are not familiar with asynchronous operations and Promises, Mozilla have written a
```
move this into a SEEALSO block

you could also rephrase this slightly:

"If you're not familiar with asynchronous operations and Promises, check out Mozilla's `detailed guide`_."

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+    $ node --experimental-repl-await
+
+
+Use Javascript to get to get ISS telemetry data
```
```suggestion
Use JavaScript to get to get ISS telemetry data
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+Use Javascript to get to get ISS telemetry data
+===============================================
+
+You will be consuming telemetry data from `Open Notify`_, a service which provides
```
```suggestion
We can get telemetry data from `Open Notify`_, a service which provides
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+===============================================
+
+You will be consuming telemetry data from `Open Notify`_, a service which provides
+a simple API to consume data from NASA. One of these data points is the current location
```
```suggestion
a simple API to consume data from NASA. Specifically, the current location
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+
+You will be consuming telemetry data from `Open Notify`_, a service which provides
+a simple API to consume data from NASA. One of these data points is the current location
+of the International Space Station. The endpoint for these data can be found at
```
```suggestion
of the International Space Station. The endpoint for this data is
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+of the International Space Station. The endpoint for these data can be found at
+`<http://api.open-notify.org/iss-now.json>`_.
+
+Start an interactive Node session using the ``node`` command. You can read the current
```
I would drop the "using the `node` command" bit, because you've already detailed how to do this above. but, if you want to keep it, I would add the `--experimental-repl-await` again here

nomicode replied:
new para at "You"

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+`<http://api.open-notify.org/iss-now.json>`_.
+
+Start an interactive Node session using the ``node`` command. You can read the current
+position of the ISS by consuming this endpoint, using ``axios``::
```
```suggestion
position of the ISS with an HTTP GET request to this endpoint, using ``axios``::
```

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+    timestamp: 1582101199,
+    message: 'success' }
+
+The endpoint returns a JSON payload consisting of ``iss_position``, which is composed of
```
```suggestion
The endpoint returns a JSON payload, which contains an``iss_position`` object with ``latitude`` and ``longitude`` data.
``

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+The endpoint returns a JSON payload consisting of ``iss_position``, which is composed of
+``latitude`` and ``longitude``, as well as some other metadata.
+
+You can encapsulate this within a single function to return only the longitude and latitude as a `WKT`_
```
```suggestion
You can encapsulate this operation with a function that returns longitude and latitude as a `WKT`_
```

mechanomi commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+
+With the table in place, you can start recording the position of the ISS.
+
+The following command calls your ``position`` function and inserts the resulting
```
```suggestion
The following command calls the ``position`` function and inserts the resulting
```

mechanomi commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+With the table in place, you can start recording the position of the ISS.
+
+The following command calls your ``position`` function and inserts the resulting
+position into your table using an `INSERT`_ query::
```
```suggestion
into the table using an `INSERT`_ query::
```

nomicode replied:
perhaps:

"The following command calls the ``position`` function and `INSERTS`_ the result
into the ``iss_position`` table:"

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+Then finally `create a table`_ suitable for writing ISS position coordinates::
+
+    > var query = `
+    ...     CREATE TABLE iss_position (
```
perhaps we should use "iss" instead of "iss_position". keeps things simple, and allows for us to maybe expand on what data we include in the future

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+===========
+
+Now you have the basics figured out, let's automate the data collection. This will
+change your approach a little bit.
```
"This will require a change of approach."

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+Now you have the basics figured out, let's automate the data collection. This will
+change your approach a little bit.
+
+You were previously using a `client`_ to connect to and insert data into CrateDB.
```
"Previously, we used a `client`_"

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+
+You were previously using a `client`_ to connect to and insert data into CrateDB.
+However, clients are ephemeral, and once closed they need to be recreated.
+Creating a new client requires a handshake with CrateDB, and
```
looks like a hard wrapping issue here (line ends too early)

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+Creating a new client requires a handshake with CrateDB, and
+this overhead cost can be prohibitive if you are creating new clients rapidly.
+
+Instead, you can use a `connection pool`_ to manage your connections.
```
perhaps add a brief sentence to this saying how connection pools address the problem

nomicode commented on docs/generate-time-series/javascript.rst:
```diff
@@ -0,0 +1,241 @@
+
+Create a new file called ``iss-position.js``, like this:
+
+.. code-block:: javascript
```
can you add a few more comments to the code in this file?

autophagy commented:
@mechanomi Thank's for the review. I have addressed your comments.

I have a question about the use of "we" vs "you" in these tutorials. Like "You can then do this" versus "we can then do this". It's a little inconsistent. I think I tend to defer to using "you" since the user is the one doing the tutorial, we're not doing it together, and sometimes it can be a little patronising. Do we have an established house style on this?

nomicode commented:
@autophagy we don't have an explicitly established house style. but we do fall back to the google documentation style guide, which covers us:

https://developers.google.com/style/person

in short, you are correct here :)

nomicode reviewed (approved):
pushed some edits, updated screenshots, added map screenshot

please review. if you're happy, let's get it merged!! 👯

nomicode commented:
split out some of the code blocks so the json is getting highlighted properly (i.e., not at all). split out other blocks for consistency

autophagy commented:
The JSON highlighting isnt working for me, nor travis. Not sure why. Just included it in the js codeblock, we can figure it out after :sweat_smile:
